sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -904,34 +904,45 @@ public String toString() {
       return name;
     }
 
+    // Maintains approximate average over last 1000 elements
+    private static class MovingAvg {
+      private static final int movingAvgWindow = 1000;
+      private double avg = 0;
+      private long numUpdates = 0;
+
+      void update(double quantity) {
+        numUpdates++;
+        avg += (quantity - avg) / Math.min(movingAvgWindow, numUpdates);
```

Comment:
Yeah, approximate. It is an approximate exponential moving average with slow decay (1/1000). Works pretty ok when there aren't huge spikes (1000x or so).